modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java

```diff
@@ -440,24 +443,67 @@ public Hit getHitInfo(float x, float y) {
-                    x -= run.getWidth();
+            if (text == null || spans == null) {
+                for (int i = 0; i < runs.length; i++) {
+                    run = runs[i];
+                    if (x < run.getWidth()) break;
+                    if (i + 1 < runs.length) {
+                        if (runs[i + 1].isLinebreak()) break;
+                        x -= run.getWidth();
+                    }
+                }
+            } else {
+                for (int i = 0; i < lineIndex; i++) {
+                    for (TextRun r: lines[i].runs) {
+                        if (r.getTextSpan() != null && r.getTextSpan().getText().equals(text)) {
+                            textWidthPrevLine += r.getLength();
+                        }
+                    }
+                }
+                int prevNodeLength = 0;
+                boolean isPrevNodeExcluded = false;
+                for (TextRun r: runs) {
+                    if (!r.getTextSpan().getText().equals(text)) {
+                        prevNodeLength += r.getWidth();
+                    }
+                    if (r.getTextSpan() != null && r.getTextSpan().getText().equals(text)) {
+                        BaseBounds textBounds = new BoxBounds();
+                        getBounds(r.getTextSpan(), textBounds);
+                        if (textBounds.getMinX() == 0 && !isPrevNodeExcluded) {
+                            x -= prevNodeLength;
+                            isPrevNodeExcluded = true;
+                        }
+                        if (x > r.getWidth()) {
+                            x -= r.getWidth();
+                            relIndex += r.getLength();
+                            continue;
+                        }
+                        run = r;
+                        break;
+                    }
                 }
             }
+
             if (run != null) {
                 int[] trailing = new int[1];
-                charIndex = run.getStart() + run.getOffsetAtX(x, trailing);
+                if (text != null && spans != null) {
+                    charIndex = run.getOffsetAtX(x, trailing);
+                    charIndex += textWidthPrevLine;
+                    charIndex += relIndex;
+                } else {
+                    charIndex = run.getStart() + run.getOffsetAtX(x, trailing);
+                }
                 leading = (trailing[0] == 0);
 
                 insertionIndex = charIndex;
                 if (getText() != null && insertionIndex < getText().length) {
                     if (!leading) {
                         BreakIterator charIterator = BreakIterator.getCharacterInstance();
-                        charIterator.setText(new String(getText()));
+                        if (text != null) {
+                            charIterator.setText(text);
+                        } else {
+                            charIterator.setText(new String(getText()));
```

Comment:
getText() is possibly an expensive operation when the layout contains a lot of text.
Since we are looking only for the following "character", is it possible to optimize this and only use a subset of spans or limit the amount of text passed to the break iterator in some other way?
What do you think?

(it's likely to be out of scope for this PR)